protocol/communities/community.go

```diff
@@ -1877,45 +1890,67 @@ func (o *Community) VerifyGrantSignature(data []byte) (*protobuf.Grant, error) {
 	return grant, nil
 }
 
-func (o *Community) CanPost(pk *ecdsa.PublicKey, chatID string, messageType protobuf.ApplicationMetadataMessage_Type) (bool, error) {
-	if o.config.CommunityDescription.Chats == nil {
-		o.config.Logger.Debug("Community.CanPost: no-chats")
+func (o *Community) CanView(pk *ecdsa.PublicKey, chatID string) (bool, error) {
+	hasAccessToChat := o.hasAccessToChat(pk, chatID)
+	if !hasAccessToChat {
 		return false, nil
 	}
 
+	chat := o.config.CommunityDescription.Chats[chatID]
+	_, isChatMember := chat.Members[common.PubkeyToHex(pk)]
+
+	return isChatMember, nil
+}
+
+func (o *Community) hasAccessToChat(pk *ecdsa.PublicKey, chatID string) bool {
+	if o.config.CommunityDescription.Chats == nil {
+		o.config.Logger.Debug("Community.hasAccessToChat: no-chats")
+		return false
+	}
+
 	chat, ok := o.config.CommunityDescription.Chats[chatID]
 	if !ok {
-		o.config.Logger.Debug("Community.CanPost: no chat with id", zap.String("chat-id", chatID))
-		return false, nil
+		o.config.Logger.Debug("Community.hasAccessToChat: no chat with id", zap.String("chat-id", chatID))
+		return false
 	}
 
 	// community creator can always post, return immediately
 	if common.IsPubKeyEqual(pk, o.ControlNode()) {
-		return true, nil
+		return true
 	}
 
 	if o.isBanned(pk) {
-		o.config.Logger.Debug("Community.CanPost: user is banned", zap.String("chat-id", chatID))
-		return false, nil
+		o.config.Logger.Debug("Community.hasAccessToChat: user is banned", zap.String("chat-id", chatID))
+		return false
 	}
 
 	if o.config.CommunityDescription.Members == nil {
-		o.config.Logger.Debug("Community.CanPost: no members in org", zap.String("chat-id", chatID))
-		return false, nil
+		o.config.Logger.Debug("Community.hasAccessToChat: no members in org", zap.String("chat-id", chatID))
+		return false
 	}
 
 	// If community member, also check chat membership next
 	_, ok = o.config.CommunityDescription.Members[common.PubkeyToHex(pk)]
 	if !ok {
-		o.config.Logger.Debug("Community.CanPost: not a community member", zap.String("chat-id", chatID))
-		return false, nil
+		o.config.Logger.Debug("Community.hasAccessToChat: not a community member", zap.String("chat-id", chatID))
+		return false
 	}
 
 	if chat.Members == nil {
-		o.config.Logger.Debug("Community.CanPost: no members in chat", zap.String("chat-id", chatID))
+		o.config.Logger.Debug("Community.hasAccessToChat: no members in chat", zap.String("chat-id", chatID))
+		return false
+	}
```

Comment:
I think you forgot to check the user being part of the chat

Reply:
it's always checked outside, maybe the name is a bit a misnomer, but I only factored out the parts that are identical between the two functions, but if you notice, we always check

Reply:
We were always checking anyway `isMember`, so I moved it inside, and now `CanPost` checks `CanView` directly